tsconfig.json

```diff
@@ -0,0 +1,25 @@
+{
+  "compilerOptions": {
+    "allowJs": true,
+    "checkJs": true,
+
+    "module": "NodeNext",
+    "moduleResolution": "NodeNext",
+    "target": "ESNext",
+    "lib": ["DOM", "ESNext"],
+    "declaration": true,
+    "emitDeclarationOnly": true,
```

Comment:
@trusktr --

This PR currently compiles every `.js` file in `/src/core` and `/src/plugins` to a corresponding `.ts` file `/dist`. It is no emitting a `d.ts` file on my dev machine. This is after a clean install.

![CleanShot 2024-03-31 at 12 53 17@2x](https://github.com/docsifyjs/docsify/assets/442527/9a88d68b-e113-411b-b6b3-7bd55ac18a11)

Reply:
The screenshot shows `.d.ts` declaration files, which is the expected result. That's what makes the types available in the demo: 

https://github.com/docsifyjs/docsify-template-plain-js-type-checked/blob/main/index.js

however the demo has docsify locally symlinked, which includes the src code. The lib otherwise doesn't export anything from the index file.

I converted the PR into a draft. Perhaps first we should export `Docsify` and allow it to accept options with `new Docsify({...})` and deprecate global `$docsify` as a way to move towards a componentization, non-globals, and ESM, and then having the type defs will make a lot more sense.

Reply:
Ok here's what I'll do: I'll make the deprecation node for `$docsify` in the same PR and expose the new non-global API (with type support as above). Let me know if you prefer the change to be separate (it won't be a big additional change, you'll be able to take a look at the specific commit).